orchard_iso_pallas.py

```diff
@@ -0,0 +1,178 @@
+            Fp(0x17033d3c60c68173573b3d7f7d681310d976bbfabbc5661d4d90ab820b12320a),
+            Fp(0x40000000000000000000000000000000224698fc094cf91b992d30ecfffffde5)
+        ]
+
+        if self == Point.identity():
+            return orchard_pallas.identity()
+        else:
+            numerator_a = c[1] * self.x * self.x * self.x + c[2] * self.x * self.x + c[3] * self.x + c[4]
+            denominator_a = self.x * self.x + c[5] * self.x + c[6]
+
+            numerator_b = (c[7] * self.x * self.x * self.x + c[8] * self.x * self.x + c[9] * self.x + c[10]) * self.y
+            denominator_b = self.x * self.x * self.x + c[11] * self.x * self.x + c[12] * self.x + c[13]
+
+            return orchard_pallas.Point(numerator_a / denominator_a, numerator_b / denominator_b)
+
+    def __init__(self, x, y, is_identity=False):
+        self.x = x
+        self.y = y
+        self.is_identity = is_identity
+
+        if is_identity:
+            assert self.x == Fp.ZERO
+            assert self.y == Fp.ZERO
+        else:
+            assert self.y * self.y == self.x * self.x * self.x + PALLAS_ISO_A * self.x + PALLAS_ISO_B
+
+    def identity():
+        p = Point(Fp.ZERO, Fp.ZERO, True)
+        return p
+
+    def __neg__(self):
+        if self.is_identity:
+            return self
+        else:
+            return Point(Fp(self.x.s), -Fp(self.y.s))
+
+    def __add__(self, a):
+        if self.is_identity:
+            return a
+        elif a.is_identity:
+            return self
+        else:
+            # <https://core.ac.uk/download/pdf/10898289.pdf> section 4.1
+            (x1, y1) = (self.x, self.y)
+            (x2, y2) = (a.x, a.y)
+
+            if x1 == x2:
+                if (y1 != y2) or (y1 == Fp(0)):
+                    return Point.identity()
+                else:
+                    return self.double()
+            else:
+                λ = (y1 - y2) / (x1 - x2)
+                x3 = λ*λ - x1 - x2
+                y3 = λ*(x1 - x3) - y1
+                return Point(x3, y3)
+
+    def __sub__(self, a):
+        return (-a) + self
+
+    def double(self):
+        if self.is_identity:
+            return self
+
+        # <https://core.ac.uk/download/pdf/10898289.pdf> section 4.1
+        λ = (Fp(3) * self.x * self.x + PALLAS_ISO_A) / (self.y + self.y)
+        x3 = λ*λ - self.x - self.x
+        y3 = λ*(self.x - x3) - self.y
+        return Point(x3, y3)
+
+    def __mul__(self, s):
+        s = format(s.s, '0256b')
+        ret = self.ZERO
+        for c in s:
+            ret = ret.double()
+            if int(c):
+                ret = ret + self
+        return ret
+
+    def __bytes__(self):
+        if self.is_identity:
+            return bytes([0] * 32)
+
+        buf = bytes(self.x)
+        if self.y.s % 2 == 1:
+            buf = buf[:31] + bytes([buf[31] | (1 << 7)])
+        return buf
+
+    def __eq__(self, a):
+        if a is None:
+            return False
```

Comment:
Why is this needed (rather than just failing on `a.is_identity`)?

Reply:
```suggestion
```

Reply:
Failing on `a.is_identity` would crash the generator (as `None` doesn't have that method). It's possible to get `None` as an output of `Point.from_bytes`, so it needs to be handled in `__eq__`.